src/debugger/android/androidPlatform.ts

```diff
@@ -48,31 +51,37 @@ export class AndroidPlatform implements IAppPlatform {
             () =>
                 Q(AndroidPlatform.RUN_ANDROID_FAILURE_PATTERNS)).process(runAndroidSpawn);
 
-        return output
-            .then(() => pkg.name())
-            .then(appName => new PackageNameResolver(appName).resolvePackageName(runOptions.projectRoot))
-            .then(packageName => {
-                this.packageName = packageName;
-                return this.deviceHelper.getConnectedDevices()
-                    .then((devices: IDevice[]) => {
-                        if (devices.length > 1) {
-                            /* more than one device or emulator */
-                            this.debugTarget = this.getTargetEmulator(runOptions, devices);
-                            if (this.debugTarget) {
-                                /* Launching is needed only if we have more than one device active */
-                                return this.deviceHelper.launchApp(runOptions.projectRoot, packageName, this.debugTarget);
-                            }
-                        } else if (devices.length === 1) {
-                            this.debugTarget = devices[0].id;
-                        }
-                    });
+        return output.finally(() =>
+            this.deviceHelper.getConnectedDevices().then(devices => {
+                this.devices = devices;
+                this.debugTarget = this.getTargetEmulator(runOptions, devices);
+            }))
+            .catch(reason => {
+                if (reason.message === AndroidPlatform.MULTIPLE_DEVICES_ERROR && this.devices.length > 1 && this.debugTarget) {
+                    /* If it failed due to multiple devices, we'll apply this workaround to make it work anyways */
+                    return this.getPackageName(runOptions.projectRoot).then(packageName =>
+                        this.deviceHelper.launchApp(runOptions.projectRoot, packageName, this.debugTarget));
+                } else {
+                    return Q.reject<void>(reason);
+                }
             }).then(() =>
                 this.startMonitoringLogCat(runOptions.logCatArguments).catch(error => // The LogCatMonitor failing won't stop the debugging experience
                     Log.logWarning("Couldn't start LogCat monitor", error)));
     }
 
     public enableJSDebuggingMode(runOptions: IRunOptions): Q.Promise<void> {
-        return this.deviceHelper.reloadAppInDebugMode(runOptions.projectRoot, this.packageName, this.debugTarget);
+        return this.getPackageName(runOptions.projectRoot).then(packageName =>
+            this.deviceHelper.reloadAppInDebugMode(runOptions.projectRoot, packageName, this.debugTarget));
+    }
+
+    // Package name is only used when we have multiple devices, or we go into debugging mode, so we initialize it lazily
```

Comment:
We always go into debugging mode when we use this class, why do we need to lazy initialize this?